File: src/AdventOfCode/Puzzles/Y2022/Day07.cs
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

using System.Diagnostics;

namespace MartinCostello.AdventOfCode.Puzzles.Y2022;

/// <summary>
/// A class representing the puzzle for <c>https://adventofcode.com/2022/day/7</c>. This class cannot be inherited.
/// </summary>
[Puzzle(2022, 07, "No Space Left On Device", RequiresData = true)]
public sealed class Day07 : Puzzle
{
    /// <summary>
    /// Gets the total size of all directories with a total size of at least 100000.
    /// </summary>
    public long TotalSizeOfDirectoriesLargerThanLimit { get; private set; }

    /// <summary>
    /// Gets the total size of the smallest directory to delete that frees enough disk space.
    /// </summary>
    public long SizeOfSmallestDirectoryToDelete { get; private set; }

    /// <summary>
    /// Finds the total size of the directories with a total size of at least 100000 and
    /// the total size of the smallest directory to delete that frees enough disk space.
    /// </summary>
    /// <param name="terminalOutput">The terminal output to parse.</param>
    /// <returns>
    /// The total size of all directories with a total size of at least 100000 and
    /// the total size of the smallest directory to delete that frees enough disk space.
    /// </returns>
    public static (long TotalSizeOfDirectoriesLargerThanLimit, long SizeOfSmallestDirectoryToDelete) GetDirectorySizes(
        IList<string> terminalOutput)
    {
        const long DiskSize = 70_000_000;
        const long SizeLimit = 100_000;
        const long UpdateSize = 30_000_000;

        var directories = GetDirectories(terminalOutput.ToArray());

        long totalSizeOfDirectoriesLargerThanLimit = directories
            .Where((p) => p.TotalSize <= SizeLimit)
            .Sum((p) => p.TotalSize);

        long consumed = directories.Single((p) => p.Container is null).TotalSize;
        long freeSpace = DiskSize - consumed;
        long required = UpdateSize - freeSpace;

        long sizeOfSmallestDirectoryToDelete = directories
            .Where((p) => p.TotalSize >= required)
            .Select((p) => p.TotalSize)
            .Min();

        return (totalSizeOfDirectoriesLargerThanLimit, sizeOfSmallestDirectoryToDelete);

        static List<Directory> GetDirectories(ReadOnlySpan<string> terminalOutput)
        {
            var root = new Directory("/", null, new());
            var current = root;

            var fileSystem = new Dictionary<string, Directory>()
            {
                [root.Path] = root,
            };

            for (int i = 0; i < terminalOutput.Length; i++)
            {
                string line = terminalOutput[i];

                if (line.AsSpan().StartsWith("$ cd "))
                {
                    string name = line[5..];

                    if (name is "..")
                    {
                        current = current.Parent!;
                    }
                    else
                    {
                        string path = name == "/" ? name : current.Path.TrimEnd('/') + "/" + name;

                        if (!fileSystem.TryGetValue(path, out var directory))
                        {
                            directory = new Directory(name, current, new());
                            fileSystem[directory.Path] = directory;
                        }

                        current = directory;
                    }
                }
                else if (line.AsSpan().StartsWith("$ ls"))
                {
                    i += ParseEntries(current, terminalOutput[(i + 1)..]);

                    foreach (var child in current.Children)
                    {
                        if (child is Directory directory)
                        {
                            fileSystem[directory.Path] = directory;
                        }
                    }
                }
            }

            return [.. fileSystem.Values];
        }

        static int ParseEntries(Directory current, ReadOnlySpan<string> terminalOutput)
        {
            int count = 0;

            for (int i = 0; i < terminalOutput.Length; i++)
            {
                string item = terminalOutput[i];

                if (item.StartsWith('$'))
                {
                    count = i;
                    break;
                }
                else if (item.AsSpan().StartsWith("dir "))
                {
                    current.Children.Add(new Directory(item[4..], current, new()));
                }
                else
                {
                    (string first, string second) = item.AsPair(' ');
                    current.Children.Add(new File(second, current, Parse<long>(first)));
                }
            }

            return count;
        }
    }

    /// <inheritdoc />
    protected override async Task<PuzzleResult> SolveCoreAsync(string[] args, CancellationToken cancellationToken)
    {
        var terminalOutput = await ReadResourceAsLinesAsync(cancellationToken);

        (TotalSizeOfDirectoriesLargerThanLimit, SizeOfSmallestDirectoryToDelete) = GetDirectorySizes(terminalOutput);

        if (Verbose)
        {
            Logger.WriteLine(
                "The sum of the total sizes of the directories with a total size of at most 100,000 is {0}.",
                TotalSizeOfDirectoriesLargerThanLimit);

            Logger.WriteLine(
                "The sizes of the directory would free up enough space on the filesystem to run the update is {0}.",
                SizeOfSmallestDirectoryToDelete);
        }

        return PuzzleResult.Create(TotalSizeOfDirectoriesLargerThanLimit, SizeOfSmallestDirectoryToDelete);
    }

    private abstract record FileSystemEntry(string Name, FileSystemEntry? Container)
    {
        public string Path =>
            Container is null ?
            Name :
            Container.Path.TrimEnd('/') + "/" + Name;
    }

    private sealed record Directory(string Name, Directory? Parent, List<FileSystemEntry> Children)
        : FileSystemEntry(Name, Parent)
    {
        public long TotalSize
        {
            get
            {
                long size = 0;

                foreach (var child in Children)
                {
                    size += child switch
                    {
                        Directory directory => directory.TotalSize,
                        File file => file.Size,
                        _ => throw new UnreachableException(),
                    };
                }

                return size;
            }
        }
    }

    private sealed record File(string Name, Directory Directory, long Size)
        : FileSystemEntry(Name, Directory);
}

File: src/AdventOfCode.Site/scripts/ts/puzzles/2022/Day07.ts
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { from } from 'linq-to-typescript';
import { Puzzle } from '../Puzzle';
import { Puzzle2022 } from './Puzzle2022';

export class Day07 extends Puzzle2022 {
    totalSizeOfDirectoriesLargerThanLimit: number;
    sizeOfSmallestDirectoryToDelete: number;

    override get name() {
        return 'No Space Left On Device';
    }

    override get day() {
        return 7;
    }

    static getDirectorySizes(
        terminalOutput: string[]
    ): [totalSizeOfDirectoriesLargerThanLimit: number, sizeOfSmallestDirectoryToDelete: number] {
        const parseEntries = (current: Directory, terminalOutput: string[]) => {
            let count = 0;

            for (let i = 0; i < terminalOutput.length; i++) {
                const item = terminalOutput[i];

                if (item.startsWith('$')) {
                    count = i;
                    break;
                } else if (item.startsWith('dir ')) {
                    current.children.push(new Directory(item.slice(4), current));
                } else {
                    const split = item.split(' ');
                    current.children.push(new File(split[1], current, Puzzle.parse(split[0])));
                }
            }

            return count;
        };

        const getDirectories = (terminalOutput: string[]) => {
            const root = new Directory('/');
            let current = root;

            const fileSystem = new Map<string, Directory>();
            fileSystem.set(root.path, root);

            for (let i = 0; i < terminalOutput.length; i++) {
                const line = terminalOutput[i];

                if (line.startsWith('$ cd ')) {
                    const name = line.slice(5);

                    if (name === '..') {
                        current = current.parent;
                    } else {
                        let path;

                        if (name === '/') {
                            path = name;
                        } else {
                            path = current.path.slice(0, -1) + '/' + name;
                        }

                        let directory: Directory;

                        if (fileSystem.has(path)) {
                            directory = fileSystem.get(path);
                        } else {
                            directory = new Directory(name, current);
                            fileSystem.set(directory.path, directory);
                        }

                        current = directory;
                    }
                } else if (line.startsWith('$ ls')) {
                    i += parseEntries(current, terminalOutput.slice(i + 1));
                    for (const child of current.children) {
                        if (child instanceof Directory) {
                            fileSystem.set(child.path, child);
                        }
                    }
                }
            }

            const result: Directory[] = [];

            for (const directory of fileSystem.values()) {
                result.push(directory);
            }

            return result;
        };

        const diskSize = 70000000;
        const sizeLimit = 100000;
        const updateSize = 30000000;

        const directories = getDirectories(terminalOutput);

        const totalSizeOfDirectoriesLargerThanLimit = directories
            .filter((p) => p.totalSize <= sizeLimit)
            .map((p) => p.totalSize)
            .reduce((x, y) => x + y, 0);

        const consumed = directories.find((p) => p.container === null).totalSize;
        const freeSpace = diskSize - consumed;
        const required = updateSize - freeSpace;

        const sizeOfSmallestDirectoryToDelete = directories
            .filter((p) => p.totalSize >= required)
            .map((p) => p.totalSize)
            .sort((x, y) => x - y)[0];

        return [totalSizeOfDirectoriesLargerThanLimit, sizeOfSmallestDirectoryToDelete];
    }

    override solveCore(_: string[]) {
        const terminalOutput = this.readResourceAsLines();

        [this.totalSizeOfDirectoriesLargerThanLimit, this.sizeOfSmallestDirectoryToDelete] = Day07.getDirectorySizes(terminalOutput);

        console.info(
            `The sum of the total sizes of the directories with a total size of at most 100,000 is ${this.totalSizeOfDirectoriesLargerThanLimit}.`
        );
        console.info(
            `The sizes of the directory would free up enough space on the filesystem to run the update is ${this.sizeOfSmallestDirectoryToDelete}.`
        );

        return this.createResult([this.totalSizeOfDirectoriesLargerThanLimit, this.sizeOfSmallestDirectoryToDelete]);
    }
}

abstract class FileSystemEntry {
    readonly path: string;

    constructor(readonly name: string, readonly container: FileSystemEntry | null) {
        if (this.container === null) {
            this.path = this.name;
        } else {
            this.path = this.container.path.slice(0, -1) + '/' + this.name;
        }
    }
}

class Directory extends FileSystemEntry {
    constructor(name: string, readonly parent: Directory | null = null, readonly children: FileSystemEntry[] = []) {
        super(name, parent);
    }

    get totalSize() {
        let size = 0;

        for (const child of this.children) {
            if (child instanceof Directory) {
                size += child.totalSize;
            } else if (child instanceof File) {
                size += child.size;
            }
        }

        return size;
    }
}

class File extends FileSystemEntry {
    constructor(name: string, directory: Directory, readonly size: number) {
        super(name, directory);
    }
}
